exporter/sentryexporter/sentry_exporter.go

```diff
@@ -4,511 +4,772 @@
-	var opBuilder strings.Builder
-	var dBuilder strings.Builder
-
-	// Generating span descriptors operates under the assumption that only one of the conventions are present.
-	// In the possible case that multiple convention attributes are available, conventions are selected based
-	// on what is most likely and what is most useful (ex. http is prioritized over FaaS)
-
-	// If http.method exists, this is an http request span.
-	if httpMethod, ok := attrs.Get(string(conventions.HTTPMethodKey)); ok {
-		opBuilder.WriteString("http")
-
-		switch spanKind {
-		case ptrace.SpanKindClient:
-			opBuilder.WriteString(".client")
-		case ptrace.SpanKindServer:
-			opBuilder.WriteString(".server")
-		case ptrace.SpanKindUnspecified:
-		case ptrace.SpanKindInternal:
-			opBuilder.WriteString(".internal")
-		case ptrace.SpanKindProducer:
-			opBuilder.WriteString(".producer")
-		case ptrace.SpanKindConsumer:
-			opBuilder.WriteString(".consumer")
+	for _, project := range projects {
+		if s.defaultTeamSlug == "" && len(project.Teams) > 0 {
+			s.defaultTeamSlug = project.Teams[0].Slug
 		}
 
-		// Ex. description="GET /api/users/{user_id}".
-		fmt.Fprintf(&dBuilder, "%s %s", httpMethod.Str(), name)
+		if _, ok := s.projectToEndpoint[project.Slug]; ok {
+			continue
+		}
 
-		return opBuilder.String(), dBuilder.String()
-	}
+		projectID := parseProjectID(project.ID)
+		if projectID <= 0 {
+			s.telemetrySettings.Logger.Warn("Skipping project due to invalid project ID",
+				zap.String("project", project.Slug))
+			continue
+		}
 
-	// If db.type exists then this is a database call span.
-	if _, ok := attrs.Get(string(conventions.DBSystemKey)); ok {
-		opBuilder.WriteString("db")
+		key, ok := projectKeyByID[projectID]
+		if !ok {
+			continue
+		}
 
-		// Use DB statement (Ex "SELECT * FROM table") if possible as description.
-		if statement, okInst := attrs.Get(string(conventions.DBStatementKey)); okInst {
-			dBuilder.WriteString(statement.Str())
-		} else {
-			dBuilder.WriteString(name)
+		dsn, err := key.DSN.ParsePublicDSN()
+		if err != nil {
+			s.telemetrySettings.Logger.Warn("Failed to parse DSN for project",
+				zap.String("project", project.Slug),
+				zap.Error(err))
+			continue
 		}
 
-		return opBuilder.String(), dBuilder.String()
+		s.projectToEndpoint[project.Slug] = &otlpEndpoints{
+			TracesURL: fmt.Sprintf("%s://%s/api/%s/integration/otlp/v1/traces/", dsn.GetScheme(), dsn.GetHost(), dsn.GetProjectID()),
+			LogsURL:   fmt.Sprintf("%s://%s/api/%s/integration/otlp/v1/logs/", dsn.GetScheme(), dsn.GetHost(), dsn.GetProjectID()),
+			PublicKey: key.Public,
+		}
 	}
+}
+
+// Shutdown stops the exporter
+func (s *endpointState) Shutdown(_ context.Context) error {
+	var shutdownErr error
+	s.shutdownOnce.Do(func() {
+		s.closing.Store(true)
+		if s.workerCancel != nil {
+			s.workerCancel()
+		}
+
+		close(s.projectCreationQueue)
+		s.workerWg.Wait()
```

Comment:
I think this can hang indefinitely, we probably want to just shutdown after a timeout if this wait blocks.